### DISCUSSION TITLE
Include custom arguments on help

### DISCUSSION THREAD

**Discussion by BlythMeister** [cake-build/cake#3284] [Q&A]
The help function shows all the targets which is great.

Is there a way (by default, or by custom in my frosting project) to also make this help show all my custom arguments that my build can take?

Currently, I maintain a MD file in the frosting repo, but would be lovely if I could include somehow.

**Comment by augustoproiete** [comment]
There's nothing built-in at the moment.

Have you tried the [Cake.ArgumentBinder](https://github.com/cake-contrib/Cake.ArgumentBinder) addin? It might give you what you're looking for, by printing out the tasks and all the argumentss  via `--showdescription`.

There's also a [feature suggestion](https://github.com/cake-build/cake/issues/2331) by @andycmaj to add a man-page style dump of all Arguments and their descriptions in a future version of Cake, which needs more discussion and design.

**Reply by BlythMeister** [reply]
argument binder looks interesting.
I have arguments that are a bit more...generic than binding to a task, mine are global and shared for all tasks.
the feature suggestion seems a bit closer to what i would be wanted to use.

Is it possible to provide some additional text which is shown as part of the help.  Because this would work for me too.
So a generic "append this to the help" type overide :D

**Reply by xforever1313** [reply]
The classes that have arguments bounded to them in Cake.ArgumentBinder dont have to be bound to a single task.  It is possible to have a class be shared between multiple tasks, or even have tasks create multiple config classes.

In a build.cake file, you could do this (there may be syntax errors, going by memory):
```c#
public class SomeConfig1
{
    [StringArgument( "somearg" )]
    public string SomeArg{ get; set }
}

public class SomeConfig2
{
    [StringArgument( "anotherarg" )]
    public string AnotherArg { get; set; }
}

SomeConfig1 config1 = CreateFromArguments<SomeConfig1>();
SomeConfig2 config2 = CreateFromArguments<SomeConfig2>();

string fullDescrption = ArgumentBinder.GetDescription<SomeConfig1>( "SomeConfig1 Arguments:") + Environment.NewLines + ArgumentBinder.GetDescription<SomeConfig2>( "SomeConfig2 Arguments:" );

Task( "task1" )
.Does(
() =>
{
    Information( config1.SomeArg );
    Information( config2.AnotherArg );
}
).Description( fullDescrption  );

Task( "task2" )
.Does(
() =>
{
    Information( config1.SomeArg );
    Information( config2.AnotherArg );
}
).Description( fullDescrption );
```

Meanwhile, in Frosting, you could do something like this to have a configuration shared between multiple tasks:

```C#
public class CustomContext: FrostingContext
{
        public CustomContext( ICakeContext context ) :
            base( context )
        {
            Config1 = context.CreateFromArguments<SomeConfig1>();
        }

        public SomeConfig1 Config1 { get; set; }
}
```

And when creating a new CakeHost, call UseContext\<CustomContext\>();

The problem is, you won't be able to get custom help text to work in Frosting.  Frosting Descriptions must be compile-time constants.  Which means you can't use ArgumentBinder's built in GetDescription function :/.
